README.rst
==============
saferedisqueue
==============

A small wrapper around `Redis <http://www.redis.io>`_ (using the "standard"
`Python redis client lib <https://pypi.python.org/pypi/redis>`_) that provides
access to a FIFO queue and allows upstream code to mark pop'ed items as
processed successfully (ACK'ing) or unsucessfully (FAIL'ing).

Failed items are automatically requeued. Addionally a backup is kept for items
that were neither ack'ed nor fail'ed, i.e. in case the consumer crashes. The
backup items will be requeued as soon as one of the consumer(s) comes up
again.

All actions are atomic and it is safe to have multiple producers and consumers
accessing the same queue concurrently.


Version compatibility with redis.py
------------------------------------

===============      ===============
redis.py             saferedisqueue
===============      ===============
2.4.10 - 2.6.x       1.x
2.7.0 - 2.7.5        no compatible version
2.7.6 - 2.10.x       2.x-3.x
===============      ===============


Usage as a library
==================

>>> queue = SafeRedisQueue(name='test')
>>> queue.put("Hello World")
'595d43b2-2e49-4e96-a1d2-0848d1c7f0d3'
>>> queue.put("Foo bar")
'1df060eb-b578-499d-bede-20db9da8184e'
>>> queue.get()
('595d43b2-2e49-4e96-a1d2-0848d1c7f0d3', 'Hello World')
>>> queue.get()
('1df060eb-b578-499d-bede-20db9da8184e', 'Foo bar')

Note: to be compatible with previous versions, 2 aliases ``push/pop`` exist. Start using the new ``put/get`` terminology as soon as possible since ``push/pop`` will be deleted in a future version.


ACK'ing and FAIL'ing
--------------------

>>> queue.put("Good stuff")
>>> queue.put("Bad stuff")
>>> uid_good, payload_good = queue.get()
>>> uid_bad, payload_bad = queue.get()
...
# process the payloads...
...
>>> queue.ack(uid_good)  # done with that one
>>> queue.fail(uid_bad)  # something didn't work out with that one, let's requeue
>>> uid, payload = queue.get()  # get again; we get the requeued payload again
>>> uid == uid_bad
True
...
# try again
...
>>> queue.ack(uid)  # now it worked; ACK the stuff now


get timeout
-----------

`SafeRedisQueue.get` accepts a timeout parameter:

- 0 (the default) blocks forever, waiting for items
- a positive number blocks that amount of seconds
- a negative timeout disables blocking


Constructor parameters
----------------------

`SafeRedisQueue` accepts ``*args, **kwargs`` and passes them to
`redis.StrictRedis`, so use whatever you need.

*Three exceptions*, use these in the keyword arguments to configure
`SafeRedisQueue` itself:

`url`
    Shortcut to use instead of a host/port/db/password combinations.
    Accepts "redis URLs" just as the redis library does, for example:

    - redis://[:password]@localhost:6379/0
    - unix://[:password]@/path/to/socket.sock?db=0

    When using this keyword parameter, all positional arguments (usually
    one the host) are ignored. Those two are equivalent:

    - ``SafeRedisQueue('localhost', port=6379, db=7)``
    - ``SafeRedisQueue(url='redis://localhost:6379/7')``

`name`
    A prefix used for the keys in Redis. Default: "0", which creates the
    following keys in your Redis DB:

    - srq:0:items
    - srq:0:queue
    - srq:0:ackbuf
    - srq:0:ackbuf_aging

    .. attention::

        The internal data structure changed in version 4.0. Pure
        producer instances (i.e. those only using `put`) are backwards-
        compatible, but consumer instances are not.

`autoclean_interval`

    An interval in seconds (default: 60) at which unconfirmed items
    (neither ACK'ed nor FAIL'ed) are requeued automatically. They are
    moved from the internal ackbuf/ackbuf_aging data structures to the
    queue again.

    Using this spawns a thread that does the autocleaning on a regular
    interval.

    Pass ``None`` to disable autocleaning. It's enabled by default!

    It's recommended to pass ``None`` when creating an instances that
    will only be used as a producer, i.e. one that will only use
    `put`/`push`.

    *The mechanism was completely re-done in version 4.0. See CHANGES
    for details.*

`serializer`
    An optional serializer to use on the items. Default: None

    Feel free to write your own serializer. It only requires a ``dumps``
    and ``loads`` methods. Modules like ``pickle``, ``json``,
    ``simplejson`` can be used out of the box.

    Note however, that when using Python 3 the ``json`` module has to be
    wrapped as it by default does not handle the ``bytes`` properly that
    is emitted by the underlying redis.py networking code. This should
    work::

        class MyJSONSerializer(object):
            @staticmethod
            def loads(bytes):
                return json.loads(bytes.decode('utf-8'))

            @staticmethod
            def dumps(data):
                return json.dumps(data)

        queue = SafeRedisQueue(name='foobar',serializer=MyJSONSerializer)

    *Added in version 3.0.0*


Command line usage
==================

For quick'n'dirty testing, you can use the script from the command line to put stuff into the queue::

    $ echo "Hello World" | python saferedisqueue.py producer

...and get it out again::

    $ python saferedisqueue.py consumer
    cbdabbc8-1c0f-4eb0-8733-fdb62a9c0fa6 Hello World


CHANGES.rst
=======
Changes
=======

4.0b0 – 2015-03-09
------------------

- BREAKING CHANGE: the data structures used on the Redis server side
  changed. While pure consumer instances (i.e. those only using `put`)
  are backwards compatibility, consuming instances are **not**. This
  means multiple consumers working on one queue must be compatible.
  Before upgrading, make sure the queue is emptied.

- The automatic requeue mechanism was reimplemented: Instead of a
  complicated algorithm that used locks and was so poorly implemented
  that it still had race conditions and caused inconsistent server-side
  state occasionally, we are now using a sorted set (ZSET) that keeps
  track of unconfirmed (neither ACK'ed *nor* FAIL'ed) and aging items.

  Note however: the new mechanism uses a thread to do the cleaning every
  x seconds (via the `autoclean_interval` parameter) independently of
  any `get`/`pop` calls. If threading gives you a headache, check out
  the code before using it. Remember, autocleaning is enabled by
  default! To disable it, pass ``autoclean_interval=None`` to the
  constructor -- this also disables any threading code; it's recommended
  to use it always for producer-only instances.


3.0.0 - 2014-11-13
------------------

- BREAKING CHANGE (very unlikely, though): `push`/`put` now returns the
  uid of the just added item. Before it always returned `None`.
- `push` has been renamed to `put`; `push` stays as an alias, but is
  deprecated and will be removed in version 4.
- `pop` has been renamed to `get`; `pop` stays as an alias, but is
  deprecated and will be removed in version 4.
- Increased version compatibility with redis.py to "<2.11".
- Python 3.4 support
- More tests. Introduced tox for testing.
- Added `serializer` parameter to support custom serializers,
  e.g. for automatic JSON conversion. See README for details, especially
  when using Python 3.


2.0.0 - 2014-06-26
------------------

- SafeRedisQueue is now officially compatible with recent versions
  (roughly speaking 2.7-2.10) of redis.py.

  For versions 2.4 and 2.6 please continue using the 1.x development
  line.

  See README.rst for details on compatibility.


1.2.0 - 2014-06-26
------------------

- Raise compatible redis.py version up to 2.6.x. Updated README with
  compatibility details.



1.1.0 - 2014-06-20
------------------

- The constructor now accepts a "url" keyword parameter that supports
  typical redis URL a la "redis://[:password]@localhost:6379/0". See
  README for details.


1.0.1 - 2013-06-26
------------------

- Changed dependency on redis to require at least version 2.4.10 as
  redis.StrictRedis, which we use, was only introduced in that version.
  This should not affect anyone negatively as you wouldn't have to be able
  to use saferedisqueue at all if your project or package used an older
  version so far.
  (Thanks Telofy!)


1.0.0 - 2013-06-05
------------------

- Released as open-source under 3-clause BSD license. Code identical to 0.3.0.


0.3.0 - 2012-05-22
------------------

- The constructor now accepts an "autoclean_interval" value to set the interval
  at which the ackbuf -> backup rotation and backup requeuing happens.
  Setting the value to `None` disables autoclean completely.
  Default: 60 (seconds).


0.2.2 - 2012-03-29
------------------

- Negative timeout makes .pop() non-blocking.


0.2.1 - 2012-03-09
------------------

- Added setup.py and publish it on our internal package directory.


0.2.0 - 2012-03-08
------------------

- Renamed methods ("push_item" -> "push" etc.)
- New autoclean method that is called on every .pop()
- Internal: New names for keys in the redis db.


test_saferedisqueue.py
# Python 2 backwards compatiblity
import sys
if sys.version_info[0] < 3:
    u = lambda x: x.decode()
    b = lambda x: x
    unichr = unichr
    basestring = basestring
    unicode = unicode
    bytes = str
    nativestr = lambda x: \
        x if isinstance(x, str) else x.encode('utf-8', 'replace')
else:
    u = lambda x: x
    unichr = chr
    unicode = str
    bytes = bytes
    b = lambda x: x.encode('latin-1') if not isinstance(x, bytes) else x
    nativestr = lambda x: \
        x if isinstance(x, str) else x.decode('utf-8', 'replace')


from uuid import uuid1

import mock
import pytest

from saferedisqueue import SafeRedisQueue


def test_put_returns_uid_string():
    queue = SafeRedisQueue(
        name='saferedisqueue-test-%s' % uuid1().hex,
        autoclean_interval=1)
    # TODO: mock uuid.uuid4 here
    uid = queue.put('blub')
    assert len(uid) == 36
    assert type(uid) is str


def test_put_internally_converts_uuid_to_str():
    queue = SafeRedisQueue(
        name='saferedisqueue-test-%s' % uuid1().hex,
        autoclean_interval=1)

    # mock it
    queue._redis = mock.Mock()
    queue._redis.pipeline.return_value = pipeline_mock = mock.Mock()
    pipeline_mock.hset.return_value = hset_mock = mock.Mock()

    uid = queue.put('blub')
    pipeline_mock.hset.assert_called_with(queue.ITEMS_KEY, uid, 'blub')
    hset_mock.lpush.assert_called_with(queue.QUEUE_KEY, uid)


def test_decode_responses_true():
    queue = SafeRedisQueue(
        name='saferedisqueue-test-%s' % uuid1().hex,
        decode_responses=True)
    unicode_string = unichr(3456) + u('abcd') + unichr(3421)
    queue.put(unicode_string)
    return_val = queue.get()[1]
    assert isinstance(return_val, unicode)
    assert unicode_string == return_val


def test_decode_responses_false():
    queue = SafeRedisQueue(name='saferedisqueue-test-%s' % uuid1().hex)
    unicode_string = unichr(3456) + u('abcd') + unichr(3421)
    queue.put(unicode_string)
    return_val = queue.get()[1]
    assert isinstance(return_val, bytes)
    assert nativestr(unicode_string) == nativestr(return_val)


# Serializer tests
# Try with json and pickle serializer

try:
    import simplejson as json
except ImportError:
    import json as json

import pickle as PickleSerializer


class MyJSONSerializer(object):
    @staticmethod
    def loads(bytes):
        return json.loads(bytes.decode('utf-8'))

    @staticmethod
    def dumps(data):
        return json.dumps(data)


@pytest.mark.parametrize("serializer", [
    MyJSONSerializer,
    PickleSerializer,
])
def test_serializer(serializer):

    queue = SafeRedisQueue(
        name='saferedisqueue-test-%s' % uuid1().hex,
        autoclean_interval=None,
        serializer=serializer
    )

    item = {'test': 'good', 'values': ['a', 'b', 'c']}

    # Test when there is an element
    queue.push(item)
    uid_item, payload_item = queue.pop()
    assert type(item) == type(payload_item)
    assert item == payload_item

    # Test when there is no element and it times out
    uid_item, payload_item = queue.pop(timeout=1)
    assert None == uid_item
    assert None == payload_item


def test_serializer_calls():
    serializer_mock = mock.Mock()
    serializer_mock.dumps.return_value = '{"dumps": "return"}'
    serializer_mock.loads.return_value = {"loads": "return"}

    queue = SafeRedisQueue(
        name='saferedisqueue-test-%s' % uuid1().hex,
        autoclean_interval=None,
        serializer=serializer_mock
    )

    item = {'test': 'good', 'values': ['a', 'b', 'c']}
    queue.push(item)
    serializer_mock.dumps.assert_called_with(item)

    assert queue.pop()[1] == {"loads": "return"}
    serializer_mock.loads.assert_called_with(b('{"dumps": "return"}'))


def test_autoclean_while_waiting():
    """If items is not acked/failed, it will appear again in the queue."""
    queue = SafeRedisQueue(
        name='saferedisqueue-test-%s' % uuid1().hex,
        autoclean_interval=0.1,
    )

    item = "example payload"
    queue.push(item)
    uid1, item1 = queue.pop()
    assert nativestr(item1) == nativestr(item)
    # While waiting here the item should be requeued and we should get
    # it again.
    uid2, item2 = queue.pop(timeout=0)
    assert uid2 == uid1
    assert nativestr(item2) == nativestr(item1)


saferedisqueue.py
# -*- coding: utf-8 -*-
"""
saferedisqueue
~~~~~~~~~~~~~~

A small wrapper around Redis that provides access to a FIFO queue and allows
upstream code to mark pop'ed items as processed successfully (ACK'ing) or
unsucessfully (FAIL'ing).

:copyright: (c) Fabian Neumann, ferret go GmbH
:license: BSD, see LICENSE for details
"""

from threading import Thread
import time
import uuid

import redis


class SafeRedisQueue(object):

    AUTOCLEAN_INTERVAL = 60

    def __init__(self, *args, **kw):
        prefix = 'srq:%s' % kw.pop('name', '0')
        self.QUEUE_KEY = '%s:queue' % prefix
        self.ITEMS_KEY = '%s:items' % prefix
        self.ACKBUF_KEY = '%s:ackbuf' % prefix
        self.ACKBUF_AGING = '%s:ackbuf_aging' % prefix
        self.AUTOCLEAN_INTERVAL = kw.pop('autoclean_interval',
                                         self.AUTOCLEAN_INTERVAL)
        self.serializer = kw.pop('serializer', None)

        url = kw.pop('url', None)
        if url is not None:
            self._redis = redis.StrictRedis.from_url(url, **kw)
        else:
            self._redis = redis.StrictRedis(*args, **kw)

        self._redis_ackbuf_to_aging = self._redis.register_script("""
            local ackbuf = KEYS[1]
            local ackbuf_aging = KEYS[2]
            local now_micros = ARGV[1]
            local ackbuflen = redis.call('llen', ackbuf)
            local iterations = math.min(ackbuflen, 100)
            for i=1,iterations,1
            do
                redis.pcall('zadd', ackbuf_aging, now_micros + i, redis.pcall('rpop', ackbuf))
            end
            return ackbuflen - iterations
            """)

        self._redis_zset_to_list_if_older = self._redis.register_script("""
            local zset = KEYS[1]
            local list = KEYS[2]
            local max_age_s = ARGV[1]
            local now_micros = ARGV[2]
            local max_score = now_micros - (max_age_s * 10^6)
            local values = redis.call('ZRANGEBYSCORE', zset, '-inf', max_score, 'LIMIT', 0, 100)
            if #values > 0
            then
                redis.call('LPUSH', list, unpack(values))
                redis.call('ZREM', zset, unpack(values))
                return redis.call('ZCOUNT', zset, '-inf', max_score)
            else
                return 0
            end
            """)

        def autoclean_loop(srq):
            interval = min(max(float(srq.AUTOCLEAN_INTERVAL) * 0.5, 0.1), 5)
            while True:
                srq._autoclean(max_seconds=interval * 0.9)
                time.sleep(interval)

        if self.AUTOCLEAN_INTERVAL is not None:
            self._autoclean_thread = Thread(
                target=autoclean_loop, name="autocleaner", args=(self,))
            self._autoclean_thread.daemon = True
            self._autoclean_thread.start()

    def _autoclean(self, max_seconds):
        deadline = time.time() + max_seconds
        while True:
            now = time.time()
            if now >= deadline:
                break
            now_micros = int(now * 10**6)
            ackbuf_remaining = self._redis_ackbuf_to_aging(
                    keys=[self.ACKBUF_KEY, self.ACKBUF_AGING],
                    args=[now_micros])
            aa_remaining = self._redis_zset_to_list_if_older(
                    keys=[self.ACKBUF_AGING, self.QUEUE_KEY],
                    args=[self.AUTOCLEAN_INTERVAL, now_micros])
            if ackbuf_remaining + aa_remaining == 0:
                break

    def put(self, item):
        """Adds an item to the queue.

        Generates a uid for later referral.
        Enqueues the uid and stores the item.
        """
        uid = str(uuid.uuid4())

        if self.serializer is not None:
            item = self.serializer.dumps(item)

        self._redis.pipeline()\
                .hset(self.ITEMS_KEY, uid, item)\
                .lpush(self.QUEUE_KEY, uid)\
                .execute()
        return uid

    # Kept for backwards compatibility
    push = put

    def get(self, timeout=0):
        """Return next item from queue. Blocking by default.

        Blocks if queue is empty, see `timeout` parameter.

        Internally this also pops uid from queue and writes it to
        ackbuffer.

        :param timeout: blocking timeout in seconds
                        - 0: block forever (default)
                        - negative: disable blocking

        """
        if timeout < 0:
            uid = self._redis.rpoplpush(self.QUEUE_KEY, self.ACKBUF_KEY)
        else:
            uid = self._redis.brpoplpush(self.QUEUE_KEY, self.ACKBUF_KEY, timeout)
        item = self._redis.hget(self.ITEMS_KEY, uid)

        # Deserialize only if the item exists: it is equal to None if it times out
        if self.serializer is not None and item is not None:
            item = self.serializer.loads(item)

        return uid, item

    # Kept for backwards compatibility
    pop = get

    def ack(self, uid):
        """Acknowledge item as successfully consumed.

        Removes uid from ackbuffer and deletes the corresponding item.
        """
        self._redis.pipeline()\
                   .lrem(self.ACKBUF_KEY, 0, uid)\
                   .zrem(self.ACKBUF_AGING, uid)\
                   .hdel(self.ITEMS_KEY, uid)\
                   .execute()

    def fail(self, uid):
        """Report item as not successfully consumed.

        Removes uid from ackbuffer and re-enqueues it.
        """
        self._redis.pipeline()\
                   .lrem(self.ACKBUF_KEY, 0, uid)\
                   .zrem(self.ACKBUF_AGING, uid)\
                   .lpush(self.QUEUE_KEY, uid)\
                   .execute()


if __name__ == "__main__":
    import sys

    def _usage():
        sys.stdout.writelines([
            'Commandline usage: python saferedisqueue.py <consumer | producer | demo>\n'
            'Example:\n'
            '    $ echo "Hello World" | python saferedisqueue.py producer\n'
            '    $ python saferedisqueue.py consumer\n'
            '    cbdabbc8-1c0f-4eb0-8733-fdb62a9c0fa6 Hello World\n'
        ])
        sys.exit(1)

    if len(sys.argv) != 2:
        _usage()

    queue = SafeRedisQueue(name='test')

    if sys.argv[1] == 'producer':
        for line in sys.stdin.readlines():
            queue.put(line.strip())
    elif sys.argv[1] == 'consumer':
        while True:
            uid, item = queue.get()
            queue.ack(uid)
            print(uid, item)
    elif sys.argv[1] == 'demo':
        map(queue.put, ['Hello', 'World'])
        while True:
            uid, item = queue.get(timeout=1)
            if uid is None:
                sys.exit()
            queue.ack(uid)
            print(uid, item)
    else:
        _usage()
